### PR TITLE
Better unboxing strategy

### PR DESCRIPTION
**Pull request by alainfrisch**
This PR changes the criterion for deciding whether to unbox a local let binding `let x = e1 in e2`.

Previously, the criterion was based purely on the shape of `e1`.  Unboxing is performed if evaluating this expression necessarily forces some boxing, i.e. if all branches naturally return an unboxed number (which then needs to be boxed to be stored in `x`).  The same criterion is applied when `x` is a mutable variable, obtained by eliminating a local reference.  This strategy leads to two kinds of "bad situations":
- A reference whose initial expression cannot be determined to be an unboxable number (at the level of Clambda) will not be unboxed, which means that all assignment to it will require boxing (if the value is obtained from an operation giving an unboxed number).  This can be seen easily with `let x = ref nan in ...` (`nan` is obtained by reading an external symbol, the information about its type is lost), or `let x = ref (f ()) in ...`.  This can be quite bad and the workaround is to make it explicit that the value is a float through `let x = ref (0. +. f ()) in ...` which is ugly.
- When `e1` is something like `if econd then ethen else eelse` then both `ethen` else `eelse` must be able to naturally return an unboxed number in order for the whole let-binding to be unboxed.  For instance, if `eelse` is `nan` (or a function call, or accessing a function's parameter, etc), then the result of `ethen` ends up being boxed.  This is for a safety reason: because of GADTs, multiple branches could return values of different types, so one cannot unbox a branch only because another one returns, say, a float.  (See http://caml.inria.fr/mantis/view.php?id=7023 )

The current PR changes the criterion as follows:
- When the let binding corresponds to an eliminated reference, and the reference was statically on an unboxable type, we always unbox the identifier, even if this does not eliminate an allocation for the initial value itself (we will potentially avoid allocations on assignments).
- Otherwise, if the let-bound identifier is known to be of an unboxable type, we decide to unbox if at least one branch would otherwise need to box its result.  This is safe, I think, because if the result type of the branching construct is a basic type, then all branches must return the same basic type (even in presence of GADTs).

The testsuite has been extended with two tests illustrating how the two kinds of "bad situations" now become boxing free.

This is the first time where unboxing depends on typing information.  This means that we won't benefit from these improvements if the type becomes known only after inlining.

### CODE REVIEW

**Comment by bobzhang** [comment]
would you take this opportunity to use record instead of tuples in the lambda types?

**Comment by alainfrisch** [comment]
> would you take this opportunity to use record instead of tuples in the lambda types?

Do you mean inline records for constructor arguments?  I don't think this is really related to this topic, so I'd better leave that for another proposal.

**Comment by alainfrisch** [comment]
Rebased to trunk.  The plumbing for extra information required for the improved unboxing strategy does not go through flambda yet.

**Comment by alainfrisch** [comment]
@chambart ported the approach of the original version to flambda: https://github.com/chambart/ocaml-1/pull/13/files

Pierre: would you feel like rebasing your work on this PR?

Everyone: I'd like to merge this soon enough since it is likely to conflict often with other changes (since it adds fields to two constructors in both lambda and clambda).  I don't think the changes are dangerous, and Pierre's original port shows that there is no blocking point to porting the approach to flambda.  So: is anyone opposed to merging this?

**Comment by alainfrisch** [comment]
The failed check is the new test, in flambda mode, as expected.  Is there an easy way to disable some tests in the suite when flambda is on?

File: bytecomp/lambda.mli [hunk 1]
```diff
@@ -152,6 +152,9 @@ and comparison =
 and array_kind =
     Pgenarray | Paddrarray | Pintarray | Pfloatarray
 
+and block_kind =
+    Pgenblock | Pfloatblock | Pboxedintblock of boxed_integer
```
Comment (chambart) [inline bytecomp/lambda.mli]: The name of the type and constructor is quite misleading as this is applied to things that are not blocs

**Comment by chambart** [comment]
My version is a bit more precise on the block representation: the fields are annotated separately with their type. Would you object to a PR on this branch to get a bit closer to that ?

By the way, I will continue to only annotate let_mutable right now. the rest might be for later.

File: bytecomp/lambda.mli (hunk 1 above)
Comment (alainfrisch) [inline bytecomp/lambda.mli]: Indeed, I'll do some renaming.

**Comment by alainfrisch** [comment]
> Would you object to a PR on this branch to get a bit closer to that ?

Sure.  Perhaps it's best to wait for the renaming to avoid conflicts. (EDIT: Renaming done.)

**Comment by alainfrisch** [comment]
Thanks @chambart !

Does anybody else  ( @xavierleroy, @mshinwell? )  want to review the whole PR?

**Comment by alainfrisch** [comment]
@chambart : can you have a look at the tests which fail with flambda (tests/translprim/comparison_table.ml.reference, tests/translprim/ref_spec.ml.reference, tests/letrec/evaluation_order_1.ml)?

**Comment by alainfrisch** [comment]
Thanks @chambart .  Can you say a few words about why the flambda changes on this PR had an impact on the evaluation order of letrec?

**Comment by mshinwell** [comment]
I will try to have a look at it, but it will be next week (w/c 21st March).

**Comment by mshinwell** [comment]
(In fact I would definitely like to review this, having just had a quick scan of it.)

**Comment by alainfrisch** [comment]
@mshinwell Have you had a chance to review this PR?

**Comment by alainfrisch** [comment]
@mshinwell Ping?

**Comment by alainfrisch** [comment]
All tests went fine; two core devs worked on it; plenty of time before the release than will include the change.  So I went ahead and merged the branch.

**Comment by gasche** [comment]
[MPR#7227](http://caml.inria.fr/mantis/view.php?id=7227) may be a regression caused by this branch -- but it is hard to tell as little information is available, besides a `tests/float-unboxing` failure message:

> Fatal error: exception Failure("float and int64 record; alloc = 56000")

**Comment by alainfrisch** [comment]
This is specific to the flambda pipeline.  @chambart Can you have a look?

**Pull request merged**